### PR TITLE
merge ui-refactor with roman-numerals

### PR DESCRIPTION
**Pull request by 1aurend**
### New File Structure
Components are now organized into three folders:

1. Views

  * Anything that deals with layouts, svgs, css, etc.

2. Logic

  * Anything with state, handles user input, etc.

3. Actions

  * Things in here probably shouldn't be components at all. These components perform actions that make most sense as actions in a reducer updating the global session state. Moving to this sort of architecture is a longer term project.

Uncategorized components are ones I have not yet refactored. (This isn't to say that all the filed ones are finished.)

There is also a folder for hooks that contains a custom hook for responsive styling.

### Completed
* Added custom hook to get rid of duplicate logic for responsive styling
* Refactored means and chart data calculations in functional style and fix the useEffect logic in those components (`Tally.js` and `ChartData.js`)
  * This fixed the bug the student found where the charts froze after 5 or so rounds.
  * Also wrote some basic tests for these new functions
* Broke the progress charts page out into several components and optimize for shorter render time
  * Was able to cut the render time in half with `memo` in `Chart.js` and separating that memoized value from any hooks.
  * Still have questions re: whether using memo on anything else actually helps. See note in `Chart.js`.
  * Also want to try implementing suspense and using loading as a fallback to see if that enables a loading screen in the delay after clicking the back button from the round stats screen.
  * Also want to see about getting away from Victory and using just HTML Canvas or something that might be faster for these basic graphs
* Got rid of all hardcoded info re: question types so the UI can accept any new questions from the generator

### In Progress
* Get rid of inline styles anywhere it makes sense in favor of styled-components
* Make any changes to the start menu logic to accommodate changes in the generator and not duplicate logic there
* Refactor `Quiz.js` and modularize the logic there better

### To Do
* Synthesize responsive styles. Lots of duplicates. Can we pass down styled components rather than js objects?
* SVG button highlights on hover-- do we like those?
* Add this to every layout:
```js
useEffect(() => {
    window.scrollTo(0, 0)
})
```
* Break `Stats.js` and `Session.js` into logic and views and modularize those calculations
* Revisit `Chord.js` which should get renamed `Vexflow` and clean up
  * This depends on what happens with the generator too; some things that are vexflow specific might get pushed down here too
* Rethink the charts and what to display if we have a larger number of question types. Should we display best and worst question type and overall? And then have some other screen with detailed numbers? Or detailed numbers below?
  * KP is also thinking about this
* Update `keystrokes.js` to accomodate new question types
* Learn how to do snapshot testing! Testing by playing is getting way too time consuming!
* **Data & DB:** Decide what info we'd want to save to a db (whether or not we hook one up now) and revamp the `Session` object.
  * This will facilitate the project of moving to a redux-like architecture where all updates to this session "store" go through a reducer.

### CODE REVIEW

File: src/components/views/QuizQuestion.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+} from 'shards-react'
+import useResponsiveStyles from '../../hooks/useResponsiveStyles'
+
+
+
+export default function QuizQuestion(props) {
+  const sizedStyles = useResponsiveStyles()
+  const { borderRadius, questionText } = sizedStyles
+  const { chord, question, colors, handleClick, onKeyPressed, currentInput } = props
+  const noteColors = (() => {
```
Comment (jsbean) [inline src/components/views/QuizQuestion.js]: Would this be any more concise as:

⚠️ Untested js code 😬 

```Javascript
const noteColor = (() => {
    return colors
        .filter(input => input.color === 'green')
        .map(input => input.input)
)()
```

If this is valid (which I am not sure if it is!), it seems like we could get rid of a logical branch, here.

Comment (jsbean) [reply src/components/views/QuizQuestion.js]: (Also, ignore any unidiomatic formatting…)

File: src/components/logic/StartMenu.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  const [session, updateSession] = useContext(Session)
+  const numQs = useRef(5)
+  const [ready, launchQuiz] = useState(false)
+  const [quiz, setQuiz] = useState()
+  const [options, updateOptions] = useState({
+    chordTypes: {triads:true, sevenths:true},
+    roots: {common:true, any:false}
+  })
+
+  // QUESTION: Does this need 'e'? Does it ultimately need to await?
```
Comment (jsbean) [inline src/components/logic/StartMenu.js]: What is `'e'` ?

**Comment by jsbean** [comment]
@1aurend, I've just got two tiny questions in the code, but it all looks good to me! There's a bunch I don't quite grok yet, but I'll get there.

Did you want to merge this before completing the "in-progress" parts and/or "todos", or do you want to bundle those all up in one PR?

File: src/components/logic/StartMenu.js (hunk 2 above)
Comment (1aurend) [reply src/components/logic/StartMenu.js]: oops... got rid of it and forgot to delete the comment!

File: src/components/views/QuizQuestion.js (hunk 1 above)
Comment (1aurend) [reply src/components/views/QuizQuestion.js]: I think that works.  If the syntax isn't valid, certainly something with that logic is. 

Fwiw, I threw this in quickly as a placeholder when I refactored `Quiz.js`. I'm not sure it'll stay. Originally there were 3 state variables, `noteColors, green, red`, and I realized (a) that `noteColors` and `green` were the same array and (b) that the state update logic back in `Quiz.js` was a mess. Basically, you have to update a state variable (or a couple other things) to trigger a re-render, which I needed after every input to show the red or green button color. I realized it made a lot more sense to have a single colors object that updates on each click, but I haven't yet finished the ripple effects of that refactor down the tree.

File: src/components/logic/Quiz.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        answer.current = {
+          answer: data[roundData.current.length].questions[chord.current.questions.length].answers[subQ.current.answers.length],
+          tries: [],
+          startTime: Date.now(),
+          endTime: '',
+          elapsedTime: '',
+        }
+        return setNextView(input)
+      default:
+        return setColors([...colors, {'input': input, color: 'red'}])
```
Comment (1aurend) [inline src/components/logic/Quiz.js]: @jsbean here's where the new consolidated `colors` object/logic is working if you didn't see it before

**Comment by 1aurend** [comment]
> @1aurend, I've just got two tiny questions in the code, but it all looks good to me! There's a bunch I don't quite grok yet, but I'll get there.
> 
> Did you want to merge this before completing the "in-progress" parts and/or "todos", or do you want to bundle those all up in one PR?

I think we should wait a bit to merge. I doubt I'll get to everything before we need to deploy, but I'd like to do a bit more, especially on breaking up logic and views before we merge.

File: src/components/views/Chart.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
+import React, { useMemo } from 'react'
```
Comment (1aurend) [inline src/components/views/Chart.js]: @jsbean this is the component that uses Victory to draw the charts if you're interested in looking into it. My sense is we should switch over to just drawing a custom chart on an html canvas or at least using vanilla d3 for the long term. And we'll need a whole bunch of new designs to accommodate all the possible question sets anyhow.

**Pull request merged**